### DISCUSSION TITLE
Losing Reactivity When I Import Solid Components in Lib

### DISCUSSION THREAD

**Discussion by captDaylight** [solidjs/solid#1706] [Q&A] [answered]
I made a reproduction of a problem I'm having when I import components as a lib from another package in my monorepo.
https://github.com/captDaylight/solid-web-component

There are two packages, `app` and `solid-components`. I'm trying to import the latter (which has two components `Counter` and `Dialog`) into `app`. 

The `Counter` component works as expected, all of the interaction is happening inside of the component. However, the `Dialog` component isn't reacting to the `open()` signal that I'm passing to it... or rather it's setting to the initial value but not reacting when I use `setOpen(true)`.

To be sure that the code I'm using is correct, I have the exact same `App.tsx` logic in the `solid-components`. In there, everything is working as expected.

**Comment by atk** [comment, marked as answer]
For the components, you need to define solid-js as a peer dependency and a "solid" export condition that points to the unfiltered JSX.

**Comment by captDaylight** [comment]
Thanks @atk for getting me unstuck!

For others, I'll expand what I did based on the answer above (the [repo](https://github.com/captDaylight/solid-web-component) is updated with this as well):

To define solid-js as a peer dependency is pretty straightforward, I just moved `solid-js` from `dependency` to `peerDependency` in `package.json`:

``` json
  "peerDependencies": {
    "solid-js": "^1.6.10"
  }
```

As for adding a "solid" export condition that points to the unfiltered JSX I updated my exports in `package.json`:

``` json
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.umd.js",
      "solid": "./dist/index.d.ts"
    }
  },
```

BUT as it was, I only had `index.mjs` and `index.umd.js` being built. So, I basically had to copy my `src` typescript into the dist folder alongside the other files. Here's my `vite.config.ts` showing the files getting copied (plus some other small changes):

``` ts
import { defineConfig } from "vite";
import solidPlugin from "vite-plugin-solid";
import path from "node:path";
import dts from "vite-plugin-dts";
import copy from "rollup-plugin-copy";

export default defineConfig(({ mode }) => {
  const isProduction = mode === "production";

  return {
    plugins: [
      solidPlugin(),
      dts({
        insertTypesEntry: true,
      }),
      isProduction &&
        copy({ // copying src into dist
          targets: [{ src: "src", dest: "dist" }],
        }),
    ].filter(Boolean),
    build: {
      lib: {
        entry: path.resolve(__dirname, "src/index.ts"),
        name: "components",
        fileName: "index",
      },
      target: "esnext",
      rollupOptions: {
        external: ["solid-js"],
      },
    },
  };
});
```

**Comment by captDaylight** [comment]
@atk are you able to explain the mechanism at work here? why was it not working before and now it is? Thanks

**Reply by atk** [reply]
The solid build plugins attempt to insert a "solid" condition into the resolver, so whenever possible, these imports get precedence over both server and browser code.

If under these conditions, solid finds JSX code, it's compiled together with the app in the same scope as the app, so it can use the same solid version, which is crucial, as the reactive code of one version will not work in the scope of another one.
